src/common/ad_mem.c

```diff
@@ -3,5 +3,1917 @@
+	}
+	arena_reorder_grp(arena, grp, grp_at, true);
+	group_decref(grp);
+
+	arena_addref(arena);
+	act->ra_arena = arena;
+	return addr;
+}
+
+static int
+arena_tx_free_addr(struct ad_arena *arena, daos_off_t addr, struct ad_tx *tx)
+{
+	struct ad_group	     *grp;
+	int		      rc;
+
+	grp = arena_locate_grp(arena, addr);
+	if (grp)
+		return -DER_NOMEM; /* the only possible failure */
+
+	rc = group_tx_free_addr(grp, addr, tx);
+	group_decref(grp);
+	return rc;
+}
+
+/**
+ * Reserve storage space with specified @size, the space should be allocated from
+ * default arena if @arena_id is set to zero, otherwise it is allocated from the
+ * provided arena.
+ */
+daos_off_t
+blob_reserve_addr(struct ad_blob *blob, int type, daos_size_t size,
+		  uint32_t *arena_id, struct ad_reserv_act *act)
+{
+	struct ad_arena	*arena = NULL;
+	daos_off_t	 addr;
+	uint32_t	 id;
+	bool		 tried;
+	int		 rc;
+
+	if (arena_id && *arena_id != AD_ARENA_ANY)
+		id = *arena_id;
+	else
+		id = blob->bb_arena_last[type]; /* the last used arena */
+
+	/* arena zero should have been created while initializing the blob */
+	D_ASSERT(id != AD_ARENA_ANY);
+	D_DEBUG(DB_TRACE, "Loading arena=%u\n", id);
+
+	rc = arena_load(blob, id, &arena);
+	if (rc) {
+		D_DEBUG(DB_TRACE, "Failed to load arena %u: %d\n", id, rc);
+		/* fall through and create a new one */
+
+	} else if (arena->ar_full) {
+		D_DEBUG(DB_TRACE, "Arena %u is full, create a new one\n", id);
+		arena_decref(arena);
+		arena = NULL;
+	}
+
+	tried = false;
+	while (1) {
+		if (arena == NULL) {
+			rc = arena_reserve(blob, type, NULL, &arena);
+			if (rc) {
+				D_ERROR("Failed to reserve new arena: %d\n", rc);
+				return rc;
+			}
+			tried = true;
+		}
+
+		D_DEBUG(DB_TRACE, "reserve from arena==%d\n", arena->ar_df->ad_id);
+		addr = arena_reserve_addr(arena, size, act);
+		if (!addr) { /* full arena */
+			D_DEBUG(DB_TRACE, "full arena=%d\n", arena->ar_df->ad_id);
+			arena_decref(arena);
+			if (tried)
+				return 0;
+
+			arena = NULL;
+			continue;
+		}
+
+		/* completed */
+		if (arena_id)
+			*arena_id = arena->ar_df->ad_id;
+
+		arena_decref(arena);
+		return addr;
+	}
+}
+
+daos_off_t
+ad_reserve(struct ad_blob_handle bh, int type, daos_size_t size,
+	   uint32_t *arena_id, struct ad_reserv_act *act)
+{
+	return blob_reserve_addr(bh.bh_blob, type, size, arena_id, act);
+}
+
+int
+tx_complete(struct ad_tx *tx, int err)
+{
+	struct ad_blob	   *blob  = tx->tx_blob;
+	struct umem_store  *store = &blob->bb_store;
+	struct ad_arena    *arena;
+	struct ad_group	   *group;
+	struct ad_free_act *fac;
+	int		    i;
+	int		    rc;
+
+	if (!err)
+		rc = store->stor_ops->so_wal_submit(store, tx->tx_id, &tx->tx_redo);
+	else
+		rc = err;
+
+	while ((arena = d_list_pop_entry(&tx->tx_ar_pub, struct ad_arena, ar_link))) {
+		arena->ar_publishing = false;
+		if (err) { /* keep the refcount and pin it */
+			d_list_add(&arena->ar_link, &blob->bb_ars_rsv);
+			continue;
+		}
+		clrbit(blob->bb_bmap_rsv, arena->ar_df->ad_id);
```

Comment:
No, only one transaction can publish an arena/group, it will be move back to the global list if it failed/canceled. The next transaction will do it again.

Reply:
Can multiple transactions reserve same arena? I mean if tx A reserved arena 1 and yield on wal commit, can tx B reserve the same arena?

Reply:
ad_blob_df::bd_bmap is set so next reservation of other tx will not select that bit?

Reply:
Oh, right. Hmm, then I'm wondering why do we need reserve bit for arena?